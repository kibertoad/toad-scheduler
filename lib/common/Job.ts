export abstract class Job {
  public readonly id: string | undefined

  protected constructor(id?: string) {
    this.id = id
  }

  abstract start(): void
  abstract stop(): void
}
