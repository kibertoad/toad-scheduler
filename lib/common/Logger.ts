export function defaultErrorHandler(id: string) {
  return (err: Error): void => {
    console.error(`Error while handling task ${id}: ${err.message}`)
  }
}

export function loggingErrorHandler(originalError: Error) {
  return (err: Error): void => {
    console.error(`Error while trying to log an error: ${err.message}`)
    console.error(`Original error: ${originalError.message}`)
  }
}
