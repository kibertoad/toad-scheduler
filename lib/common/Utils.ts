export function isPromise(value: any): value is Promise<any> {
  return value && value.then !== undefined ? true : false
}
