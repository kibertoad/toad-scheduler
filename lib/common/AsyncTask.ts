import { defaultErrorHandler, loggingErrorHandler } from './Logger'
import { isPromise } from './Utils'

export class AsyncTask {
  public isExecuting: boolean
  private readonly id: string
  private readonly handler: () => Promise<void>
  private readonly errorHandler: (err: Error) => void | Promise<void>

  constructor(id: string, handler: () => Promise<void>, errorHandler?: (err: Error) => void) {
    this.id = id
    this.handler = handler
    this.errorHandler = errorHandler || defaultErrorHandler(this.id)
    this.isExecuting = false
  }

  execute(): void {
    this.isExecuting = true
    this.handler()
      .catch((err: Error) => {
        const errorHandleResult = this.errorHandler(err)
        if (isPromise(errorHandleResult)) {
          // If we fail while handling an error, oh well
          errorHandleResult.catch(loggingErrorHandler(err))
        }
      })
      .finally(() => {
        this.isExecuting = false
      })
  }
}
