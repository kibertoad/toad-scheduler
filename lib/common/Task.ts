import { defaultErrorHandler, loggingErrorHandler } from './Logger'
import { isPromise } from './Utils'

export class Task {
  public isExecuting: boolean
  private readonly id: string
  private readonly handler: () => void
  private readonly errorHandler: (err: Error) => void | Promise<void>

  constructor(id: string, handler: () => void, errorHandler?: (err: Error) => void) {
    this.id = id
    this.handler = handler
    this.errorHandler = errorHandler || defaultErrorHandler(this.id)
    this.isExecuting = false
  }

  execute(): void {
    this.isExecuting = true
    try {
      this.handler()
    } catch (err: any) {
      const errorHandleResult = this.errorHandler(err)
      if (isPromise(errorHandleResult)) {
        // If we fail while handling an error, oh well
        errorHandleResult.catch(loggingErrorHandler(err))
      }
    }
    this.isExecuting = false
  }
}
