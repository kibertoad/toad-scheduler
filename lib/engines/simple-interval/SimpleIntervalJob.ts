import Timeout = NodeJS.Timeout
import { Job, JobStatus } from '../../common/Job'
import { SimpleIntervalSchedule, toMsecs } from './SimpleIntervalSchedule'
import { Task } from '../../common/Task'
import { AsyncTask } from '../../common/AsyncTask'

export class SimpleIntervalJob extends Job {
  private timer?: Timeout
  private readonly schedule: SimpleIntervalSchedule
  private readonly task: Task | AsyncTask

  constructor(schedule: SimpleIntervalSchedule, task: Task | AsyncTask, id?: string) {
    super(id)
    this.schedule = schedule
    this.task = task
  }

  start(): void {
    const time = toMsecs(this.schedule)
    // See https://github.com/kibertoad/toad-scheduler/issues/24
    if (time >= 2147483647) {
      throw new Error(
        'Due to setInterval limitations, no intervals longer than 24.85 days can be scheduled correctly. toad-scheduler will eventually include a workaround for this, but for now your schedule is likely to break.'
      )
    }

    // Avoid starting duplicates and leaking previous timers
    if (this.timer) {
      this.stop()
    }

    this.timer = setInterval(() => {
      this.task.execute()
    }, time)
  }

  stop(): void {
    if (!this.timer) {
      return
    }
    clearInterval(this.timer)
    this.timer = undefined
  }

  getStatus(): JobStatus {
    if (this.timer) {
      return JobStatus.RUNNING
    }
    return JobStatus.STOPPED
  }
}
