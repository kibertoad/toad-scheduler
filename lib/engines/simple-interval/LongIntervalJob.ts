import Timeout = NodeJS.Timeout
import { AsyncTask } from '../../common/AsyncTask'
import { Job, JobStatus } from '../../common/Job'
import { Task } from '../../common/Task'
import { JobOptions, SimpleIntervalJob } from './SimpleIntervalJob'
import { SimpleIntervalSchedule, toMsecs } from './SimpleIntervalSchedule'

const MAX_TIMEOUT_DURATION_MS = 2147483647

export class LongIntervalJob extends Job {
  private childJob?: SimpleIntervalJob
  private timer?: Timeout
  private readonly schedule: SimpleIntervalSchedule
  private readonly task: Task | AsyncTask
  private readonly preventOverrun: boolean
  constructor(schedule: SimpleIntervalSchedule, task: Task | AsyncTask, options: JobOptions = {}) {
    super(options.id)
    this.preventOverrun = options.preventOverrun ?? true
    this.schedule = schedule
    this.task = task

    // Create a future date for this task and get the number of ms
    const taskPeriod = toMsecs(schedule)
    // Initiate time-eating logic
    if (taskPeriod >= MAX_TIMEOUT_DURATION_MS) {
      this.setTimeEatingJob(taskPeriod)
    }
  }

  private setTimeEatingJob(taskPeriod: number): void {
    const mainTaskExecutionDate = new Date()
    mainTaskExecutionDate.setTime(Date.now() + taskPeriod)
    const mainTaskExecutionTime = mainTaskExecutionDate.getTime()
    const startingRemainingMs = mainTaskExecutionTime - Date.now()

    const timeEater = new Task('time eating task', () => {
      const remainingMs = mainTaskExecutionTime - Date.now()
      if (remainingMs >= MAX_TIMEOUT_DURATION_MS) {
        /* istanbul ignore next */
        this.childJob?.stop()
        this.childJob = new SimpleIntervalJob(
          {
            milliseconds: Math.min(MAX_TIMEOUT_DURATION_MS - 1, remainingMs),
          },
          timeEater
        )
        this.childJob.start()
      } else {
        /* istanbul ignore next */
        this.childJob?.stop()
        this.childJob = new SimpleIntervalJob(
          {
            milliseconds: Math.min(MAX_TIMEOUT_DURATION_MS - 1, remainingMs),
          },
          new Task('Final mile task', () => {
            this.setTimeEatingJob(toMsecs(this.schedule))
            return this.task.execute()
          })
        )
        this.childJob.start()
      }
    })

    this.childJob?.stop()
    this.childJob = new SimpleIntervalJob(
      {
        milliseconds: Math.min(MAX_TIMEOUT_DURATION_MS - 1, startingRemainingMs),
      },
      timeEater
    )
    this.childJob.start()
  }

  start(): void {
    if (this.childJob) {
      return this.childJob.start()
    }

    const time = toMsecs(this.schedule)
    // Avoid starting duplicates and leaking previous timers
    if (this.timer) {
      this.stop()
    }

    if (this.schedule.runImmediately) {
      this.task.execute()
    }

    this.timer = setInterval(() => {
      if (!this.task.isExecuting || !this.preventOverrun) {
        this.task.execute()
      }
    }, time)
  }

  stop(): void {
    if (this.childJob) {
      return this.childJob.stop()
    }
    if (!this.timer) {
      return
    }
    clearInterval(this.timer)
    this.timer = undefined
  }

  getStatus(): JobStatus {
    if (this.childJob) {
      return this.childJob.getStatus()
    }
    if (this.timer) {
      return JobStatus.RUNNING
    }
    return JobStatus.STOPPED
  }
}
