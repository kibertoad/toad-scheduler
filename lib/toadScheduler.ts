import { SimpleIntervalEngine } from './engines/simple-interval/SimpleIntervalEngine'
import { SimpleIntervalJob } from './engines/simple-interval/SimpleIntervalJob'
import { Job, JobStatus } from './common/Job'
import { LongIntervalJob } from './engines/simple-interval/LongIntervalJob'
import { CronJob } from './engines/cron/CronJob'
import { CronJobEngine } from './engines/cron/CronJobEngine'

type EngineRegistry = {
  simpleIntervalEngine?: SimpleIntervalEngine
  cronJobEngine?: CronJobEngine
}

export class ToadScheduler {
  private readonly engines: EngineRegistry
  private readonly jobRegistry: Record<string, Job>

  constructor() {
    this.engines = {}
    this.jobRegistry = {}
  }

  /**
   * @deprecated in favor of unified addJob method
   */
  addIntervalJob(job: SimpleIntervalJob | LongIntervalJob) {
    if (!this.engines.simpleIntervalEngine) {
      this.engines.simpleIntervalEngine = new SimpleIntervalEngine()
    }

    this.registerJob(job)
    this.engines.simpleIntervalEngine.add(job)
  }

  /**
   * @deprecated in favor of unified addJob method
   */
  addLongIntervalJob(job: LongIntervalJob): void {
    return this.addIntervalJob(job)
  }

  /**
   * @deprecated in favor of unified addJob method
   */
  addSimpleIntervalJob(job: SimpleIntervalJob): void {
    return this.addIntervalJob(job)
  }

  private registerJob(job: Job): void {
    if (job.id) {
      if (this.jobRegistry[job.id]) {
        throw new Error(`Job with an id ${job.id} is already registered.`)
      }
      this.jobRegistry[job.id] = job
    }
  }

  /**
   * @deprecated in favor of unified addJob method
   */
  addCronJob(job: CronJob): void {
    if (!this.engines.cronJobEngine) {
      this.engines.cronJobEngine = new CronJobEngine()
    }

    this.registerJob(job)
    this.engines.cronJobEngine.add(job)
  }

  addJob(job: Job): void {
    if (job instanceof SimpleIntervalJob || job instanceof LongIntervalJob) {
      this.addIntervalJob(job)
    } else if (job instanceof CronJob) {
      this.addCronJob(job)
    } else {
      throw new Error('Unsupported job type.')
    }
  }

  stop(): void {
    for (const engine of Object.values(this.engines)) {
      engine?.stop()
    }
  }

  getById(id: string): Job {
    const job = this.jobRegistry[id]
    if (!job) {
      throw new Error(`Job with an id ${id} is not registered.`)
    }
    return job
  }

  existsById(id: string): boolean {
    const job = this.jobRegistry[id]
    if (!job) {
      return false
    }
    return true
  }

  removeById(id: string): Job | undefined {
    const job = this.jobRegistry[id]
    if (!job) {
      return
    }
    job.stop()
    delete this.jobRegistry[id]

    return job
  }

  stopById(id: string): void {
    const job = this.getById(id)
    job.stop()
  }

  startById(id: string): void {
    const job = this.getById(id)
    job.start()
  }

  getAllJobs(): Job[] {
    return Object.values(this.jobRegistry)
  }

  getAllJobsByStatus(status: JobStatus): Job[] {
    return Object.values(this.jobRegistry).filter((value) => {
      return value.getStatus() === status
    })
  }
}
