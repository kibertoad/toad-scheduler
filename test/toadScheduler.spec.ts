import { ToadScheduler } from '../lib/toadScheduler'
import { SimpleIntervalJob } from '../lib/engines/simple-interval/SimpleIntervalJob'
import { Task } from '../lib/common/Task'
import { NoopTask } from './utils/testTasks'
import { advanceTimersByTime, mockTimers, unMockTimers } from './utils/timerUtils'
import { JobStatus } from '../lib/common/Job'
import { expectToMatchObject } from './utils/assertUtils'

describe('ToadScheduler', () => {
  beforeEach(() => {
    mockTimers()
  })

  afterEach(() => {
    unMockTimers()
  })

  describe('getById', () => {
    it('returns job correctly', () => {
      const scheduler = new ToadScheduler()
      const task = new NoopTask()
      const job = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id' },
      )
      scheduler.addSimpleIntervalJob(job)
      const retrievedJob = scheduler.getById('id')
      expect(retrievedJob.getStatus()).toBe('running')
      expect(retrievedJob).toBe(job)
    })
  })

  describe('existsById', () => {
    it('returns true for existing job', () => {
      const scheduler = new ToadScheduler()
      const task = new NoopTask()
      const job = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id' },
      )
      scheduler.addSimpleIntervalJob(job)
      const result = scheduler.existsById('id')
      expect(result).toBe(true)
    })

    it('returns false for non-existing job', () => {
      const scheduler = new ToadScheduler()
      const task = new NoopTask()
      const job = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id' },
      )
      scheduler.addSimpleIntervalJob(job)
      const result = scheduler.existsById('id2')
      expect(result).toBe(false)
    })
  })

  describe('removeById', () => {
    it('correctly removes job by id', () => {
      let counter = 0
      let counter2 = 0
      const scheduler = new ToadScheduler()
      const task = new Task('simple task', () => {
        counter++
      })
      const task2 = new Task('simple task2', () => {
        counter2++
      })
      const job = new SimpleIntervalJob(
        {
          milliseconds: 1,
        },
        task,
        { id: 'job1' },
      )
      const job2 = new SimpleIntervalJob(
        {
          milliseconds: 10,
        },
        task2,
        { id: 'job2' },
      )

      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)

      expect(counter).toBe(0)
      expect(counter2).toBe(0)

      const deletedJob = scheduler.removeById('job2')
      expect(deletedJob?.id).toMatch('job2')
      expect(() => {
        scheduler.getById('job2')
      }).toThrowError(/not registered/)
      const nonExistingJob = scheduler.removeById('job2')
      expect(nonExistingJob).toBeUndefined()

      advanceTimersByTime(2)
      expect(counter).toBe(2)
      expect(counter2).toBe(0)
      advanceTimersByTime(10)
      expect(counter).toBe(12)
      expect(counter2).toBe(0)

      scheduler.stop()
    })
  })

  describe('stopById', () => {
    it('throws an error when non-existent id is stopped', () => {
      const scheduler = new ToadScheduler()
      expect(() => {
        scheduler.stopById('dummy')
      }).toThrowError(/Job with an id dummy is not registered./)
    })

    it('correctly stops job by id', () => {
      let counter = 0
      let counter2 = 0
      const scheduler = new ToadScheduler()
      const task = new Task('simple task', () => {
        counter++
      })
      const task2 = new Task('simple task2', () => {
        counter2++
      })
      const job = new SimpleIntervalJob(
        {
          milliseconds: 1,
        },
        task,
        { id: 'job1' },
      )
      const job2 = new SimpleIntervalJob(
        {
          milliseconds: 10,
        },
        task2,
        { id: 'job2' },
      )

      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)

      expect(counter).toBe(0)
      expect(counter2).toBe(0)

      scheduler.stopById('job2')

      advanceTimersByTime(2)
      expect(counter).toBe(2)
      expect(counter2).toBe(0)
      advanceTimersByTime(10)
      expect(counter).toBe(12)
      expect(counter2).toBe(0)

      scheduler.stop()
    })
  })

  describe('startById', () => {
    it('throws an error when non-existent id is started', () => {
      const scheduler = new ToadScheduler()
      expect(() => {
        scheduler.startById('dummy')
      }).toThrowError(/Job with an id dummy is not registered./)
    })

    it('correctly starts job by id', () => {
      let counter = 0
      let counter2 = 0
      const scheduler = new ToadScheduler()
      const task = new Task('simple task', () => {
        counter++
      })
      const task2 = new Task('simple task2', () => {
        counter2++
      })
      const job = new SimpleIntervalJob(
        {
          milliseconds: 1,
        },
        task,
        { id: 'job1' },
      )
      const job2 = new SimpleIntervalJob(
        {
          milliseconds: 10,
        },
        task2,
        { id: 'job2' },
      )

      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)

      expect(counter).toBe(0)
      expect(counter2).toBe(0)

      scheduler.stopById('job2')
      scheduler.startById('job2')

      advanceTimersByTime(2)
      expect(counter).toBe(2)
      expect(counter2).toBe(0)
      advanceTimersByTime(10)
      expect(counter).toBe(12)
      expect(counter2).toBe(1)

      scheduler.stop()
    })
  })

  describe('listJobs', () => {
    it('returns all jobs', () => {
      const scheduler = new ToadScheduler()
      const task = new NoopTask()
      const job = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id' },
      )
      const job2 = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id2' },
      )
      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)
      job.stop()

      const retrievedJobs = scheduler.getAllJobs()
      expectToMatchObject(retrievedJobs, [
        {
          id: 'id',
        },
        {
          id: 'id2',
        },
      ])
    })
  })

  describe('listJobs', () => {
    it('returns jobs filtered by status', () => {
      const scheduler = new ToadScheduler()
      const task = new NoopTask()
      const job = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id' },
      )
      const job2 = new SimpleIntervalJob(
        {
          seconds: 20,
        },
        task,
        { id: 'id2' },
      )
      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)
      job.stop()

      const retrievedJobs = scheduler.getAllJobsByStatus(JobStatus.STOPPED)
      expectToMatchObject(retrievedJobs, [
        {
          id: 'id',
        },
      ])
    })
  })

  describe('common', () => {
    it('throws an error when duplicate id is registered', () => {
      const scheduler = new ToadScheduler()
      const job = new SimpleIntervalJob(
        {
          milliseconds: 1,
        },
        new NoopTask(),
        { id: 'job1' },
      )
      const job2 = new SimpleIntervalJob(
        {
          milliseconds: 10,
        },
        new NoopTask(),
        { id: 'job1' },
      )

      scheduler.addSimpleIntervalJob(job)
      expect(() => {
        scheduler.addSimpleIntervalJob(job2)
      }).toThrowError(/Job with an id job1 is already registered/)
    })

    it('correctly stops without any jobs', () => {
      const scheduler = new ToadScheduler()
      scheduler.stop()
    })

    it('correctly stops even when some engines are not initialized', () => {
      const scheduler = new ToadScheduler()

      scheduler['engines'].simpleIntervalEngine = undefined
      scheduler.stop()
    })

    it('correctly handles multiple jobs', () => {
      let counter = 0
      let counter2 = 0
      const scheduler = new ToadScheduler()
      const task = new Task('simple task', () => {
        counter++
      })
      const task2 = new Task('simple task2', () => {
        counter2++
      })
      const job = new SimpleIntervalJob(
        {
          milliseconds: 1,
        },
        task,
      )
      const job2 = new SimpleIntervalJob(
        {
          milliseconds: 10,
        },
        task2,
      )

      scheduler.addSimpleIntervalJob(job)
      scheduler.addSimpleIntervalJob(job2)

      expect(counter).toBe(0)
      expect(counter2).toBe(0)
      advanceTimersByTime(2)
      expect(counter).toBe(2)
      expect(counter2).toBe(0)
      advanceTimersByTime(10)
      expect(counter).toBe(12)
      expect(counter2).toBe(1)

      scheduler.stop()
    })
  })
})
