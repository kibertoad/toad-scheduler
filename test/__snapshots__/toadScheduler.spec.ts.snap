// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToadScheduler listJobs returns all jobs 1`] = `
Array [
  SimpleIntervalJob {
    "id": "id",
    "preventOverrun": true,
    "schedule": Object {
      "seconds": 20,
    },
    "task": NoopTask {
      "errorHandler": [Function],
      "handler": [Function],
      "id": "dummy",
      "isExecuting": false,
    },
    "timer": undefined,
  },
  SimpleIntervalJob {
    "id": "id2",
    "preventOverrun": true,
    "schedule": Object {
      "seconds": 20,
    },
    "task": NoopTask {
      "errorHandler": [Function],
      "handler": [Function],
      "id": "dummy",
      "isExecuting": false,
    },
    "timer": Object {
      "hasRef": [Function],
      "ref": [Function],
      "refed": true,
      "refresh": [Function],
      "unref": [Function],
      Symbol(Symbol.toPrimitive): [Function],
    },
  },
]
`;

exports[`ToadScheduler listJobs returns jobs filtered by status 1`] = `
Array [
  SimpleIntervalJob {
    "id": "id",
    "preventOverrun": true,
    "schedule": Object {
      "seconds": 20,
    },
    "task": NoopTask {
      "errorHandler": [Function],
      "handler": [Function],
      "id": "dummy",
      "isExecuting": false,
    },
    "timer": undefined,
  },
]
`;
